libpages/server.go

```diff
@@ -0,0 +1,206 @@
+// Copyright 2017 Keybase Inc. All rights reserved.
+// Use of this source code is governed by a BSD
+// license that can be found in the LICENSE file.
+
+package libpages
+
+import (
+	"context"
+	"crypto/ecdsa"
+	"crypto/elliptic"
+	"crypto/rand"
+	"net/http"
+	"sync"
+	"time"
+
+	lru "github.com/hashicorp/golang-lru"
+	"github.com/keybase/kbfs/libkbfs"
+	"go.uber.org/zap"
+	"golang.org/x/crypto/acme"
+	"golang.org/x/crypto/acme/autocert"
+)
+
+// ServerConfig holds configuration parameters for Server.
+type ServerConfig struct {
+	AutoDirectHTTP     bool
+	DomainWhitelist    []string
+	UseStaging         bool
+	Logger             *zap.Logger
+	UseDiskCacheForDev bool
+}
+
+const fsCacheSize = 2 << 15
+
+// Server handles incoming HTTP requests by creating a Root for each host and
+// serving content from it.
+type Server struct {
+	config     ServerConfig
+	kbfsConfig libkbfs.Config
+
+	fsCache *lru.Cache
+
+	whiteList     map[string]bool
+	whiteListOnce sync.Once
+}
+
+func (s *Server) getFS(
+	ctx context.Context, root Root) (http.FileSystem, error) {
+	fsCached, ok := s.fsCache.Get(root)
+	if !ok {
+		fs, err := root.MakeFS(ctx, s.config.Logger, s.kbfsConfig)
+		if err != nil {
+			return nil, err
+		}
+		s.fsCache.Add(root, fs)
+		return fs, nil
+	}
+	fs, ok := fsCached.(http.FileSystem)
+	if !ok {
```

Comment:
Seems like this should log a pretty nasty warning, as it would be very unexpected.